src/main/java/org/broadinstitute/barclay/argparser/CommandLinePluginDescriptor.java

```diff
@@ -76,103 +110,127 @@
-     * help/doc generation.
+     * @return the default plugins enabled for this descriptor as a list of Object. Used for
+     * help/documentation generation for this descriptor.
+     *
+     * NOTE: calling this method after argument parsing (and thus after {@link #validateAndResolvePlugins}
+     * is called) may return a different list than calling it before parsing, since the resolution
+     * policy is implementation-dependent and may depend on the actual argument specified by the user
+     * on the command line. For example, the user may have specified that a plugin originally included
+     * in the list of defaults be disabled. The corresponding plugin class would be returned by this
+     * mehtod BEFORE parsing, but not after, since it would have been removed by
+     * {@link #validateAndResolvePlugins}.
      */
-    public abstract List<Object> getDefaultInstances();
+    public abstract List<T> getDefaultInstances();
 
     /**
-     * @return an ordered List of actual plugin instances that have been specified on the command
-     * line, in the same order they were obtained/created by {@line #getInstance}).
+     * @return an ordered list containing the plugins that result from resolving all command line
+     * arguments with any default plugins that have been provided to this descriptor. This list
+     * represents the plugins that will actually be used by the consumer. The resolution policy is
+     * descriptor-dependent.
+     *
+     * NOTE: calling this method before argument parsing (and thus before {@link #validateAndResolvePlugins}
+     * has been called) may return a different list than calling it after parsing, since the resolution
+     * policy is implementation-dependent and may depend on the actual arguments specified by the user
+     * on the command line.
      */
-    public abstract List<T> getAllInstances();
+    public abstract List<T> getResolvedInstances();
 
     /**
+     * Return the allowable values for the String argument of this plugin descriptor
+     * that is specified by {@code longArgName}. Used for help/documentation generation for this descriptor.
+     * If the value is unrecognized the implementation should throw IllegalArgumentException.
+     *
+     * @param longArgName
+     * @return Set<String> of allowable values, or empty set if any value is allowed
+     */
+    public abstract Set<String> getAllowedValuesForDescriptorHelp(String longArgName);
+
+    /**
+     * Return the class object for the plugin with simple class name {@code pluginName}
+     * Used for help/usage and documentation generation.
+     *
      * @param pluginName Name of the plugin requested
      * @return Class object for the plugin instance requested
      */
-    public abstract Class<?> getClassForInstance(final String pluginName);
+    public abstract Class<?> getClassForPluginHelp(final String pluginName);
```

Comment:
Why isn't it bounded? I still think that the methods should be bounded because the plugin descriptor should be used with classes of `T` without exception...

Reply:
See above.

Reply:
Enforce in the contract that the returned value should be lower-bounded for `T` and that it is safe to cast it.